cfgrammar/src/lib/yacc/ast.rs

```diff
@@ -5,10 +5,56 @@ use std::{
 
 use indexmap::{IndexMap, IndexSet};
 
-use super::{Precedence, YaccGrammarError, YaccGrammarErrorKind};
+use super::{parser::YaccParser, Precedence, YaccGrammarError, YaccGrammarErrorKind, YaccKind};
 
 use crate::Span;
 
+pub struct ASTValidation {
+    pub ast: GrammarAST,
+    pub errs: Vec<YaccGrammarError>,
+}
+
+/// Provides access to an ast which may or may not be valid,
+/// An AST which is invalid is provided for further analysis.
+impl ASTValidation {
+    /// The AST will be invalid if errors were encountered during
+    /// construction in which case `errors()` will be non-empty.
+    pub fn new(yacc_kind: YaccKind, s: &str) -> Self {
+        let mut errs = Vec::new();
+        let ast = match yacc_kind {
+            YaccKind::Original(_) | YaccKind::Grmtools | YaccKind::Eco => {
+                let mut yp = YaccParser::new(yacc_kind, s.to_string());
+                let _ = yp.parse().map_err(|e| errs.extend(e));
+                let mut ast = yp.ast();
+                let _ = ast.complete_and_validate().map_err(|e| errs.push(e));
+                ast
+            }
+        };
+        ASTValidation { ast, errs }
+    }
+
+    /// Returns a `GrammarAST` which may contain errors and should be validated.
+    ///
+    /// You may use an ast with errors to perform further analysis.
+    /// Consumers of `ASTValidation` such as [YaccGrammar::new_with_storaget](crate::yacc::YaccGrammar::new_with_storaget]
+    /// should propagate any errors within it along with any additional errors that they produce.
+    ///
+    /// User specified consumers should not drop the errors without propagating them upwards.
```

Comment:
I'm not quite sure I understand this sentence: they can't drop the errors can they?

Reply:
Well, they kind of can because the caller *owns* an `ASTValidation`, and so they can drop that.
I should probably just reformulate the sentance to say `should make certain to propagate the errors upwords` which is the important point.

Reply:
I half get it, but what's a "user specified consumer"?

Reply:
It would be something that calls `ASTWithValidityInfo`, that never goes on to construct a `YaccGrammar` from it.
Once you construct a `YaccGrammar`, the errors get collated with the `YaccGrammarResult`, 
But if the caller never proceeds that far, it is their own burden to propagate those errors somehow.
`user specified consumer`, should probably be replaced by `caller`, *on the other hand* if the user does construct a `YaccGrammar`, they *don't* want to do anything with the errors at risk of double reporting them.

Double reporting should be obvious once it occurs, I think I should just say here:
> `YaccGrammar` will consume the errors upon construction. If not constructing
>  a `YaccGrammar` it is the callers responsibility to report errors.

or something to that effect.

Reply:
Nope, I still don't get it :) The warning seems to suggest there's something very unusual about `ASTWithValidityInfo` that requires careful thought from the user in order for bad things not to happen, but I'm not sure I see that link. Maybe I'm missing out on something!

Reply:
I think you might get it, It certainly is unusual, I would prefer not having something like this public. But it has to be because the callers that need it are in the `lrpar` crate rather than `cfgrammar`.  The only real bright side to it is that passing it to `YaccGrammar` is one of the few reasonable things to do with it.

Reply:
Let me ask the question in the other way: what happens if you fling `ASTWithValidityInfo` away and then create a `YaccGrammar`?

Reply:
You would just not see or report any of the errors produced when creating the first `GrammarAST`, and rebuild a new `GrammarAST` when you create the `YaccGrammar`.  The worst that can really happen is that `drop(AstWithValidityInfo)` is an implicit `.ok()`.

Reply:
OK, so I *think* we should remove the "User specified consumers should not drop the errors without propagating them upwards" comment because there's nothing special about this struct?

Reply:
Right, so I think the sentence at https://github.com/softdevteam/grmtools/pull/354/commits/1e6c7d3556741f4c9dbad09dce4b5bef2f28e5b8#diff-8aeec91eb9693494ba63059be659b7c45094f8747d2b54ac288f5ec62ec56aafR27 should be deleted?

Reply:
Removed it in [4311b96](https://github.com/softdevteam/grmtools/pull/354/commits/4311b961a1ef081c7c3136786a5e3108ebc0e851)